### ISSUE TITLE
Errors thrown without NLog.Web in extensions

### ISSUE THREAD

**Issue by ShaneCourtrille** [NLog/NLog#2768]
**NLog version**: 4.5.4.7700

**Platform**: .Net 4.5

We recently upgraded to 4.5.4.7700 from 4.5.0 and have started to see errors for a target we've had setup for forever which uses aspnet-user-identity from NLog.Web.  I can resolve the error by adding NLog.Web as an extension but I'm wondering if that is expected as I can't find any mention of it elsewhere.  Also the exception that gets thrown seems wrong.

- What is the current result?

Event viewer shows the following exception which doesn't help at all and seems wrong?:

`Message: Type is not resolved for member 'NLog.NLogConfigurationException,NLog, Version=4.0.0.0, Culture=neutral, PublicKeyToken=5120e14c03d0593c'.`

NLog internal logs show:

```
2018-06-11 12:40:45.6381 Warn Error when setting property 'Body' on 'Mail Target[exchange]' Exception: System.ArgumentException: LayoutRenderer cannot be found: 'aspnet-user-identity'. Is NLog.Web not included?
2018-06-11 12:40:45.6381 Warn Error when setting '<sty... **snipped for length** on attibute 'body'
2018-06-11 12:40:45.6381 Error Parsing configuration from C:\Data\Service\NLog.config failed. Exception: NLog.NLogConfigurationException: Exception when parsing C:\Data\Service\NLog.config.  ---> NLog.NLogConfigurationException: Error when setting property 'Body' on Mail Target[exchange] ---> System.ArgumentException: LayoutRenderer cannot be found: 'aspnet-user-identity'. Is NLog.Web not included?
```

- What is the expected result?

Same behavior as 4.5.0 OR a blog post/documentation letting us know about the change.

**Comment by snakefoot** [comment]
When updating NLog. Did you update any of these:

https://www.nuget.org/packages/NLog.Extensions.Logging
https://www.nuget.org/packages/NLog.Web

Are you using ASP.NET Core?

**Comment by snakefoot** [comment]
@ShaneCourtrille  Could you try other versions of NLog (Ex 4.5.1, 4.5.2 or 4.5.3) to help find when the change was introduced ?

**Comment by snakefoot** [comment]
@ShaneCourtrille Please also enable the internal logger. And copy-paste the output on debug-level. Both when working and not working: https://github.com/NLog/NLog/wiki/Internal-Logging

**Comment by ShaneCourtrille** [comment]
Yes, I updated NLog.Web actually to latest at the same time.  

No, we're not on ASP.Net Core but I'll do some experimentation to see what I can figure out on my side for NLog/NLog.Web and I'll get those logs for you.

**Comment by snakefoot** [comment]
@ShaneCourtrille Been able to cast some light on when the behavior change was introduced?

**Comment by bosma4u** [comment]
I experienced a similar problem, NLog 4.4.13 worked fine and from 4.5 the problem manifests itself every time the application pool is restarted. When the application is published, everything is fine, but if the application pool is restarted for any reason, the logging stops working.

Error ConfigSectionHandler error. Exception: NLog.NLogConfigurationException: Exception when parsing D:\virtual_hosts\loadtest\web.config.  ---> NLog.NLogConfigurationException: Error when setting property 'Layout' on File Target[(unnamed)] ---> System.ArgumentException: LayoutRenderer cannot be found: 'aspnet-SessionId'. Is NLog.Web not included?

**Comment by snakefoot** [comment]
Could you activate the NLog internal logger. And give the log during normal startup and the log during pool restart.



Sent from my Samsung device


-------- Original message --------

**Comment by bosma4u** [comment]
There is nothing in nlog-internal after a normal start. I include the trace after an apppoolrestart:
[nlog4.5-after-apppoolrestart.log](https://github.com/NLog/NLog/files/2244299/nlog4.5-after-apppoolrestart.log)

**Comment by snakefoot** [comment]
Please configure the log level to Trace for the internal logger. And try again needing the details leading up to the errors



Sent from my Samsung device


-------- Original message --------

**Comment by bosma4u** [comment]
Here is a log with level trace.
10:35 Normal operation with nLog 4.4.13
10:38 Publish of a version with nLog 4.5.7, working
10:40 Recycle of app pool leading to the errors.
[nlog4.5-internal-trace.log](https://github.com/NLog/NLog/files/2244486/nlog4.5-internal-trace.log)

**Comment by snakefoot** [comment]
@bosma4u I'm confused how you have been able to avoid any logging about "Start auto loading" or "ScanAssembly" (For all 3 scenarios). This part is very important as it handles the automatic loading of extension-dlls like "NLog.Web" which is needed for ${aspnet-SessionId} to work.

**Comment by snakefoot** [comment]
@bosma4u I'm confused how you have been able to avoid any logging about "Start auto loading" or "ScanAssembly" (For all 3 scenarios). This part is very important as it handles the automatic loading of extension-dlls like "NLog.Web" which is needed for ${aspnet-SessionId} to work (along with other aspnet-friends).

Is the internal-logger enabled somewhere later in the process?

**Comment by bosma4u** [comment]
In web.config we include a configuration like.
<nlog>
    <include file="D:\Config\NLogLoadtest.config" />
</nlog>
Nlog.config was present though, with internalLogLevel="Off"
Here is a log with "Start auto loading" present.
[nlog4.5-internal-trace-restart.log](https://github.com/NLog/NLog/files/2248487/nlog4.5-internal-trace-restart.log)

**Comment by snakefoot** [comment]
Could you do as before. Where you had NLog 4.4 startup + Nlog 4.5 startup + Nlog 4.5 pool restart


Sent from my Samsung device


-------- Original message --------

**Comment by bosma4u** [comment]
11:53 4.4.13 deploy
12:01 4.5.7 deploy
12:02 4.5.7 restart
[nlog4.4deploy-4.5deploy-restart.log](https://github.com/NLog/NLog/files/2249039/nlog4.4deploy-4.5deploy-restart.log)

**Comment by snakefoot** [comment]
Can you show file properties for the NLog.dll deployed? (Most importantly File Description)


Sent from my Samsung device


-------- Original message --------

**Comment by bosma4u** [comment]
![image](https://user-images.githubusercontent.com/41120092/43520359-e55aae9a-9592-11e8-808f-2e22e8bd2c32.png)

**Comment by snakefoot** [comment]
@bosma4u Thank you so much for your help and patience. I have found the bug introduced with #2643 in NLog ver. 4.5.2

The problem is the last line (should instead return `assemblyLocation`)

https://github.com/NLog/NLog/blob/da2c45e6fd733d622bc0680a14a09540a1645a31/src/NLog/Config/ConfigurationItemFactory.cs#L501-L507

**Comment by snakefoot** [comment]
Created PR #2830 to resolve this bug (Use NLog-assembly location as primary search location, instead of AppDomain-BaseDirectory).

**Comment by 304NotModified** [comment]
Thanks @snakefoot !

**Comment by 304NotModified** [comment]
resolved with  #2830

**Comment by ShaneCourtrille** [comment]
Thanks so much Rolf for taking the time I couldn't find to get this
resolved!
